### PR TITLE
[Heartbeat]Adds negative body match

### PR DESCRIPTION
**Pull request by lowryxiao**
## What does this PR do?

**Adds new option "positive_check_on_http_body" under "check.response" to control if the check on http response body is positive or negative.**

## Why is it important?

**Heartbeat can match HTTP response bodies using an arbitrary regexp with check.response.body, however this only matches positively, customers might want to perform a negative match.**

## Checklist

- [x] My code follows the style guidelines of this project
- [x ] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made a corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues
- Opens #20728 

## Use cases
Add positive and negative options under check.response.body. 

Case1): If the customer specifies both of them, then the response body should match these 2 conditions at the same time,
1.  for the positive option, the positive regex pattern should match the response body;
2. for the negative option, the negative regex pattern should not match the response body;

For example:
```yaml
# Both negative and positive option gave
check.response.body:
positive:
- foo
- faa
negative:
- bar
- boo
```

In this case, check.response.body should match the following condition:
```bash
check.response.body ~ /(foo|faa)/ && check.response.body !~ /(bar|boo)/
```

Case2): If the customer specifies any of the positive/negative options, then the response body should match the specified type of check:
1.  for the positive option, the positive regex pattern should match the response body;
2. for the negative option, the negative regex pattern should not match the response body;

Case3): If the custom doesn't give any positive or negative option, we just take it as a positive check, this used to tolerate the previous version of Heartbeat.

### CODE REVIEW

**Comment by andrewvc** [comment]
Hi @lowryxiao thanks for the submission. It is very much appreciated.

I've been thinking about this issue for a bit, and I think we can improve the config syntax here. Having a sibling of a config option `positive_check_on_http_body`  that modifies the `body` attribute is confusing IMHO.

We should allow the flexibility for users to specify both positive and negative matches. Your proposed patch lets people flip the match, but I think we can do better.  My proposal is to extend the syntax such that the following syntaxes would work:


```yaml
# the original syntax
check.response.body:
- foo
- bar

# Both negative and positive
check.response.body:
positive:
  - foo
negative:
  - bar
```

This would require a fallback parsing of the config to determine which one is used

What are your thoughts? CC @shahzad31 who also chimed in on the issue.

**Comment by lowryxiao** [comment]
@andrewvc 
Sorry for the delay, still waiting for my CLA approval from my company.

Got your point, what you want to improve are:
1. add `positive` and `negative` options under `check.response.body`. If customer specifies both of them, then the response body should match these 2 regex conditions at the same time, right?

like the example you given:
# Both negative and positive
check.response.body:
  positive:
    - foo
  negative:
    - bar

So the HTTP response body(let's just call it check.response.body) should match the following regex at the same time:

check.response.body ~ /foo/ && check.response.body !~ /bar/

2. If custom doesn't give the `positive` or `negative` option, we just take it as `positive` check.

Please feel free to comment if I have misunderstanding or you have better idea.
Thanks.

**Comment by andrewvc** [comment]
@lowryxiao that sounds exactly correct! Looking forward to your contribution :)

**Comment by lowryxiao** [comment]
@andrewvc Sure, It's my pleasure to do that.
I've already pushed my latest changes to this PR, still waiting for my CLA approval.
Wondering if you could review my code changes first before the CLA? If that I will have more time to improve it if any suggestions from you.

**Comment by vjsamuel** [comment]
@andrewvc Lowry has been added to the corporate CLA and passed on to the CLA team. is there some way to verify?

**Comment by andrewvc** [comment]
Hmmm, let me see if there's a way. @elastic/observablt-robots what's going on with the CLA? How can we verify that everything is correct here?

**Comment by kuisathaverat** [comment]
I've checked the use and the PR, the user does not have the CLA signed/approve so the PR do not pass the CLA
 
Screenshot 2020-10-21 at 09 56 11

**Comment by lowryxiao** [comment]
Thank @vjsamuel @andrewvc @kuisathaverat for helping.
If I understand right, the flow should be like this:
get CLA signed for the user side first, then the PR can pass though the CLA sign, right?

But as Vijay said, I have been added to the corporate CLA and passed on to the CLA team.
So what I can do now is to wait for the approval of CLA team?

**Comment by kuisathaverat** [comment]
>If I understand right, the flow should be like this:
get CLA signed for the user side first, then the PR can pass though the CLA sign, right?
But as Vijay said, I have been added to the corporate CLA and passed on to the CLA team.
So what I can do now is to wait for the approval of CLA team?

In that case, I guess the document signed with your data is not processed yet.

**Comment by kuisathaverat** [comment]
cla/check

**Comment by andrewvc** [comment]
@lowryxiao can you double-check that you've signed the CLA? If so, let me see if I can track down what went wrong in the process. @kuisathaverat it sounds like our database doesn't have @lowryxiao in it for sure right?

**Comment by vjsamuel** [comment]
@andrewvc i see that Carl Spataro from elastic has ratified the addition of Lowry into the corporate CLA of eBay. Can Carl help?

**Comment by kuisathaverat** [comment]
>it sounds like our database doesn't have @lowryxiao in it for sure right?

@andrewvc the CLA check app does not find him, and is not in the list of users, I'll ask infra.

**Comment by kuisathaverat** [comment]
We’ve contacted the legal team and got the list of the contributors for eBay, and there is no @lowxryxiao in that list.  It’s better when they contact us again through the cla-questions@elastic.co to clarify things about the CLAs

**Comment by lowryxiao** [comment]
Thank all guys for helping me to resolve this CLA approval problem.
One thing wants to make sure, I updated my github profile from `github.com/xiaofengmanlou` to `github.com/lowryxiao` before I requested the CLA approval to my company.
Wondering if this makes us in trouble?

**Comment by kuisathaverat** [comment]
> Thank all guys for helping me to resolve this CLA approval problem.
> One thing wants to make sure, I updated my github profile from `github.com/xiaofengmanlou` to `github.com/lowryxiao` before I requested the CLA approval to my company.
> Wondering if this makes us in trouble?

I do not think so, xiaofengmanlou is not in the DB either

**Comment by olksdr** [comment]
@kuisathaverat CLA check is passing now.

**Comment by lowryxiao** [comment]
Thanks for all your help.
Eventually, we're here.

**Comment by andrewvc** [comment]
jenkins run the tests

File: heartbeat/monitors/active/http/check_test.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check_test.go:149]: Thanks for adding these tests!

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:140]: Why are `positive` and `negative` declared as vars here? It's a bit odd since they are only used once. Can we switch to just using string literals later.

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:145]: Can we combine this line and the following? That's more common in our code.

`if p, ok := b.([]interface{}); ok {` is what I'm looking for

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:139]: It'd be cleaner to just name the return values and avoid the extra var declaration here. I also wouldn't mind using longer var names for clarity. 

`func parseBody(b interface{}) (positiveMatch []match.Matcher, negativeMatch []match.Matcher) {`

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:185]: I see the intent here, using the `if/else` branches to return an optimal function, but I'm not sure the tradeoff in readability / fragility of code is worth it.

I'd prefer to either only run the final case with the two `for` loops OR to build up an array of function pointers, with each one corresponding to a positive or negative match and iterate over that.

That said, we're really getting into micro optimizations here, and I'm not convinced that will yield much in this case, and I don't think it's worth the effort to benchmark, so I think the two for loops makes the most sense.

**Review by andrewvc — CHANGES_REQUESTED** [review]
Functionally looks great. Could use some code changes. Thanks for the contribution! Needs docs as well (with an example). The docs changes would show up here: https://www.elastic.co/guide/en/beats/heartbeat/current/monitor-http-options.html#monitor-http-check

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:140]: Sure, will adjust them.

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:145]: Goot catch, thanks for your suggestion.

Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:145]: Good catch, thanks for your suggestion.

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:139]: That makes sense.

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go]: Yea, I made things complex.
Refactored this function as you suggested, please help to review.

File: heartbeat/monitors/active/http/check_test.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check_test.go:149]: Sure, add another test in case positive and negative are both empty.

Comment (lowryxiao) [reply heartbeat/monitors/active/http/check_test.go:149]: Added the doc in `heartbeat/docs/monitors/monitor-http.asciidoc` file.
Please help to review. thx.

**Comment by lowryxiao** [comment]
@andrewvc 
Updated all occurrences as you mentioned, could you help to review them again?
Thanks.

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:161]: This seems more like an error condition, and one that could be caught earlier (in the `for checkType loop`). In the case of error conditions (a type like `ppositive`) we should try and help the user out and report an error like "invalid check type 'ppositive', expected 'positive' or 'negative'.

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:184]: Why do we need to bother with the flags when we could just `return` if the positive match fails? That seems a lot simpler.

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:205]: Same goes here, why not just fail immediately and get rid of the flags?

**Review by andrewvc — CHANGES_REQUESTED** [review]
Looking very close to done, just added a few comments. Will re-run CI as well and check on that. Thanks for the contribution!

**Comment by v1v** [comment]
Jenkins run the tests please

**Comment by andrewvc** [comment]
It looks like this patch has broken a couple unit tests for large HTTP responses and redirects. Those will need to be fixed here as well @lowryxiao

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:184]: That is in case the customer only give the negative check, for this case, the positive check is always failing.

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:205]: This is for the case customer only specify the positive check, for this case, the negative check is always a success.

**Comment by lowryxiao** [comment]
Updated the design, there should be 4 cases:

> @andrewvc
> Sorry for the delay, still waiting for my CLA approval from my company.
> 
> Wondering what you recommend to improve are:
> case 1. add `positive` and `negative` options under `check.response.body`. If the customer specifies both of them, then the response body should match these 2 regex conditions at the same time, right?
> 
> like the example you given:
> 
> `# Both negative and positive`
> `check.response.body:`
> ` positive:`
> ` - foo`
> ` negative:`
> ` - bar`
> 
> So the HTTP response body(let's just call it check.response.body) should match the following regex at the same time:
> 
> `check.response.body ~ /foo/ && check.response.body !~ /bar/`
> 
> case 2. If custom doesn't give the `positive` or `negative` option, we just take it as a `positive` check.
> 
> case 3. there's only `positive` check under `check.response.body`.
> like the example:
> 
> `# Only positive`
> `check.response.body:`
> ` positive:`
> ` - foo`
> 
> The HTTP response body(let's just call it check.response.body) should match the following regex:
> 
> `check.response.body ~ /foo/`
> 
> case 4. there's only `negative` check under `check.response.body`.
> like the example:
> 
> `# Only positive`
> `check.response.body:`
> ` negative:`
> ` - bar`
> 
> The HTTP response body(let's just call it check.response.body) should match the following regex:
> 
> `check.response.body !~ /bar/`
> 
> Please feel free to comment if I have misunderstood or you have a better idea.
> Thanks.

**Comment by lowryxiao** [comment]
> It looks like this patch has broken a couple unit tests for large HTTP responses and redirects. Those will need to be fixed here as well @lowryxiao

Got it, will take care of them.

**Comment by lowryxiao** [comment]
@andrewvc Updated all the codes related to the comments as we discussed and fixed the unit tests as you mentioned.
Please have a review when you get a chance.

**Comment by andrewvc** [comment]
@lowryxiao this is looking great, and all tests are passing! I'm going to give this a final review tomorrow, but I'm expecting to be able to merge it then. Thanks for all the great work you've put into this :)

File: heartbeat/monitors/active/http/check.go
Comment (andrewvc) [inline heartbeat/monitors/active/http/check.go:167]: Love this error check

**Review by andrewvc — APPROVED** [review]
LGTM, looks great!

File: heartbeat/monitors/active/http/check.go
Comment (lowryxiao) [reply heartbeat/monitors/active/http/check.go:167]: Thanks.

**Comment by lowryxiao** [comment]
@andrewvc Thanks for merging this PR.

**Pull request merged**
